src/apps/external_executor/grpc.h

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+
+#include "ccf/odata_error.h"
+#include "ds/serialized.h"
+
+#include <arpa/inet.h>
+#include <variant>
+#include <vector>
+
+namespace ccf::grpc
+{
+  using CompressedFlag = uint8_t;
+  using MessageLength = uint32_t;
+
+  static constexpr size_t message_frame_length =
+    sizeof(CompressedFlag) + sizeof(MessageLength);
+
+  MessageLength read_message_frame(const uint8_t*& data, size_t& size)
+  {
+    auto compressed_flag = serialized::read<CompressedFlag>(data, size);
+    if (compressed_flag >= 1)
+    {
+      throw std::logic_error(fmt::format(
+        "gRPC compressed flag has unexpected value {}", compressed_flag));
+    }
+    return ntohl(serialized::read<MessageLength>(data, size));
+  }
+
+  void write_message_frame(uint8_t*& data, size_t& size, size_t message_size)
+  {
+    CompressedFlag compressed_flag = 0;
+    serialized::write(data, size, compressed_flag);
+    serialized::write(data, size, htonl(message_size));
+  }
+
+  template <typename In>
+  In get_grpc_payload(const std::shared_ptr<ccf::RpcContext>& ctx)
+  {
+    auto& request_body = ctx->get_request_body();
+    auto request_content_type =
+      ctx->get_request_header(http::headers::CONTENT_TYPE);
+
+    auto data = request_body.data();
+    auto size = request_body.size();
+
+    // Only remove gRPC frame is content type is gRPC so that it is possible to
+    // send raw protobuf payload for debug
+    if (request_content_type == http::headervalues::contenttype::GRPC)
+    {
+      auto message_length = grpc::read_message_frame(data, size);
+      if (size != message_length)
+      {
+        throw std::logic_error(fmt::format(
+          "Error in gRPC frame: frame size is {} but messages is {} bytes",
+          size,
+          message_length));
+      }
+      ctx->set_response_header(
+        http::headers::CONTENT_TYPE, http::headervalues::contenttype::GRPC);
+    }
+    In in;
+    if (!in.ParseFromArray(data, size))
+    {
+      throw std::logic_error(
+        fmt::format("Error deserialising protobuf payload of size {}", size));
+    }
+    return in;
+  }
+
+  template <typename Out>
+  void set_grpc_response(
+    const Out& resp, const std::shared_ptr<ccf::RpcContext>& ctx)
+  {
+    size_t r_size = resp.ByteSizeLong();
+    std::vector<uint8_t> r(r_size);
+
+    auto r_data = r.data();
+
+    auto request_content_type =
+      ctx->get_request_header(http::headers::CONTENT_TYPE);
+    if (request_content_type == http::headervalues::contenttype::GRPC)
+    {
+      r_size += grpc::message_frame_length;
+      r.resize(r_size);
+      r_data = r.data();
+      grpc::write_message_frame(r_data, r_size, resp.ByteSizeLong());
+      ctx->set_response_header(
+        http::headers::CONTENT_TYPE, http::headervalues::contenttype::GRPC);
+    }
```

Comment:
I'd expect a `415 Unsupported Media Type` error if this header is not `application/grpc`, as we do in the JSON handler, but I can't see that anywhere. If we want to maintain a debug path where we can send raw protobuf by hand, it should probably insist on _no_ content type, or some other dedicated debug content type.

Reply:
I'm no longer sure keeping the non-gRPC (i.e. raw protobuf) path is actually helpful for debugging now that unary gRPC works so I've removed it and now mandates that the `content-type` is gRPC (`415` is returned otherwise).